File: leetcode/maximum-value-of-k-coins-from-piles.cpp
class Solution {
public:
    int maxValueOfCoins(vector<vector<int>>& p, int k) {
        vector<int> f(k + 1);
        for (auto a : p)
        {
            for (int i = k; i >= 0; i--)
            {
                int s = 0;
                for (int j = 1; j <= a.size() && i + j <= k; j++)
                {
                    s += a[j - 1];
                    f[i + j] = max(f[i + j], f[i] + s);
                }
            }
        }
        return f.back();
    }
};

File: leetcode/maximum-value-of-k-coins-from-piles.py
class Solution:
    def maxValueOfCoins(self, p: List[List[int]], k: int) -> int:
        f = [0] * (k + 1)
        for a in p:
            g = f[:]
            s = 0
            for i in range(len(a)):
                s += a[i]
                for j in range(i + 1, k + 1):
                    g[j] = max(g[j], f[j - (i + 1)] + s)
            f = g
        return f[-1]